documentation/projects/proposals/popularity_optimizations/20230420-implementation_plan_popularity_optimizations.md

```diff
@@ -0,0 +1,488 @@
+catalog database.
+
+For simplicity, I will use `image` as an example, but note that equivalent
+structures exist for `audio` and will exist for any media types added in the
+future. Descriptions are not exhaustive; some details that are not relevant have
+been omitted for brevity.
+
+#### `image`
+
+The 'main' image table. Rows in this table may contain **raw** popularity data
+from the provider in the `meta_data` column, such as the number of times this
+image has been viewed or downloaded at the source provider.
+
+This table is updated by **ingestion**, which may:
+
+- insert newly ingested images
+- update previously ingested images (for example, by updating raw popularity
+  data)
+
+#### `image_popularity_metrics`
+
+A table which contains known 'metrics': that is, the name of the metadata field
+which contains raw popularity data for each provider. _Not all providers have a
+configured popularity metric._
+
+This table only needs to be updated when a metric is added or modified.
+
+#### `image_popularity_constants`
+
+A materialized view which calculates, for each provider that supports popularity
+data, a popularity constant that can be used to normalize raw popularity scores.
+This constant is calculated using the 85th percentile of the given metric (total
+views, etc). Because of this, the constant needs to be periodically updated for
+accuracy as more data is consumed.
+
+Refreshing this view results in the constants being recalculated. We currently
+do this once a month. It must also be done whenever a new metric is added.
+
+#### `image_view`
+
+A materialized view sourced from the `image` table, which adds the calculated
+`standardized_popularity` column. This normalized popularity score is calculated
+by a simple operation using the raw score and the popularity constant.
+
+The effect of refreshing the view is to:
+
+- Add all images which were ingested since the last time the view was refreshed,
+  and calculate their popularity
+- Update existing records which may have been modified on the `image` table
+  since the last refresh
+- Update the `standardized_popularity` for all records to use _the current
+  popularity constant_
+
+In a data refresh, the data from this view is copied into a new table in the API
+DB, which is used to build new Elasticsearch indices and eventually swapped with
+the live tables. As a result, the view **must** be refreshed before the data
+refresh can run, or newly ingested data would not make it to the API.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+The basic strategy proposed here for decoupling popularity calculations from the
+data refresh is:
+
+1. Add `standardized_popularity` as a `nullable` column on the `image` and
+   `audio` tables. ([link](#add-standardized-popularity-to-the-media-tables))
+1. Update the `provider_dag_factory` to calculate `standardized_popularity` **at
```

Comment:
I'm curious about the reference to `provider_dag_factory` here, but not `ProviderDataIngester`. Is that just because the metadata is standardized enough across providers that it'll be DRYer to do it in the dag factory instead?

Reply:
Good point, it should probably read ```Update the `provider_dag_factory` AND the `ProviderDataIngester`...```! The actual calculation will happen in the `ProviderDataIngester`, but there are also updates needed to the dag factory to pass the necessary constants into the ingester class. There's more details in the implementation section.